### PR TITLE
Add proj-data

### PR DESCRIPTION
**Pull request by xylar**
This is a package with data needed for `proj` for systems with a firewall or without an internet connection.

This replaces `datumgrid-feedstock` and the `proj-datumgrid-*` packages it produces.

See https://github.com/conda-forge/proj.4-feedstock/pull/71 for related discussion.

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml" 
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

File: recipes/proj-data/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  # Each grid has a *README.txt file with the license info, that is copied with
+  # the datum files.
+  # license_file: *README.txt
+  summary: Data for Cartographic Projections and Coordinate Transformations Library
+
+extra:
+  recipe-maintainers:
+    - ocefpaf
+    - xylar
+    - snowman2
```
Comment (xylar) [inline recipes/proj-data/meta.yaml]: @ocefpaf and @snowman2, are you willing to be maintainers? Are there others I should add?

File: recipes/proj-data/build.sh [hunk 2]
```diff
@@ -0,0 +1,4 @@
+export DATADIR="${PREFIX}/share/proj"
```
Comment (xylar) [inline recipes/proj-data/build.sh]: This should be updated related to this discussion: https://github.com/conda-forge/proj.4-feedstock/pull/71#issuecomment-596211051

**Comment by xylar** [comment]
The build is going to fail until `proj 7.0.0.` is available.

File: recipes/proj-data/build.sh (hunk 2 above)
Comment (xylar) [reply recipes/proj-data/build.sh]: I started working on this but this clearly has to be a directory within the environment so it's not clear to me that `$XDG_DATA_HOME` can be of any use in this case.

Comment (ocefpaf) [reply recipes/proj-data/build.sh]: > I started working on this but this clearly has to be a directory within the environment so it's not clear to me that `$XDG_DATA_HOME` can be of any use in this case.

Not sure what we can do then. Makes sense that `conda` won't allow us to write this at build time.

**Comment by snowman2** [comment]
This is a useful guide: https://proj.org/resource_files.html

**Comment by xylar** [comment]
Oh, dear! So is there any way to override this with an environment variable?

**Comment by xylar** [comment]
Oe will proj look in $PREFIX/share/proj first?

**Comment by xylar** [comment]
Sorry, was reading the docs on my phone.  So that seems okay.  It will look in a "PROJ user writable directory" but if nothing is found there, it will go on to look in `$PROJ_LIB`, which, for us, is `$PREFIX/share/proj`, so I think we're good.

File: recipes/proj-data/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  noarch: generic
+
+requirements:
+  run:
+    - proj >=7.0.0
+
+test:
+  commands:
+    # TODO: we need real tests for each datumgrid.
+    - echo -105 40 | cs2cs +init=epsg:4326 +to +init=epsg:2975
```
Comment (xylar) [inline recipes/proj-data/meta.yaml]: @snowman2 or @hobu, any input on what would be more meaningful tests?  I didn't see any super obvious examples in the PROJ documentation that make use of these data sets but I also have to admit that I use `proj` through `pyproj` so I don't know base package very well to begin with.

Comment (snowman2) [reply recipes/proj-data/meta.yaml]: If it helps any `cs2cs` is essentially `Transformer.from_crs`. There may be some tests in the pyproj repo and PROJ repo that might be helpful.

**Comment by xylar** [comment]
@ocefpaf and @snowman2, again wondering if you're okay with being maintainers.

File: recipes/proj-data/meta.yaml (hunk 3 above)
Comment (xylar) [reply recipes/proj-data/meta.yaml]: Thanks, I think I found a decent test in the documentation that I was able to modify for the new `proj-data` file naming and format.

File: recipes/proj-data/build.sh (hunk 2 above)
Comment (xylar) [reply recipes/proj-data/build.sh]: Right, and even if we could, it wouldn't end up in the package.

**Comment by xylar** [comment]
@conda-forge/help-python or @conda-forge/staged-recipes, I think this is ready for review.  Issues to keep in mind:
* the data has several licenses, so no one license file can be provided (similar to existing https://github.com/conda-forge/datumgrid-feedstock)
* Because this is `noarch: generic`, I didn't add a `build.bat` so the failure under Windows builds is expected.

**Comment by ocefpaf** [comment]
> @ocefpaf and @snowman2, again wondering if you're okay with being maintainers.

Sure. Count me in.

**Comment by snowman2** [comment]
I can be a maintainer as well 👍

**Comment by hobu** [comment]
Please add me as a maintainer to both this and proj

**Comment by xylar** [comment]
@hobu, you have been added.  I will make a PR now, to add you to `proj`.

**Comment by xylar** [comment]
@conda-forge/staged-recipes, all the listed maintainers have agreed to maintain this recipe.

File: recipes/proj-data/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    - test -f $PREFIX/share/proj/ca_nrc_ntv1_can.tif
+    - echo -111 50 | cs2cs +proj=latlong +ellps=clrk66 +nadgrids=ca_nrc_ntv1_can.tif +to +proj=latlong +ellps=GRS80 +datum=NAD83
+    - echo -105 40 | cs2cs +init=epsg:4326 +to +init=epsg:2975
+
+about:
+  home: https://github.com/OSGeo/proj-data
+  license: Public domain | X/MIT | BSD-2/3/4-Clause | CC0 | CC-BY (v3.0 or later) | CC-BY-SA (v3.0 or later)
+  # Each grid has a *README.txt file with the license info, that is copied with
+  # the datum files.
+  # license_file: *README.txt
```
Comment (scopatz) [inline recipes/proj-data/meta.yaml]: You have to have a license file

Comment (xylar) [reply recipes/proj-data/meta.yaml]: What do you suggest?

Comment (xylar) [reply recipes/proj-data/meta.yaml]: And why was https://github.com/conda-forge/datumgrid-feedstock (the precursor to this recipe) allowed not to have one?

**Comment by xylar** [comment]
@scopatz, I don't feel great about this but I added the MIT license from the `proj` repo (called "COPYING").  It really doesn't apply to any of the data sets packaged here, but rather to the `proj` package.  But if there's a need for a single license, I guess that's the only one I can come up with.

@ocefpaf, any thoughts?

**Comment by ocefpaf** [comment]
We decided a while ago that the licenses in the data folder was enough for
these pkgs.

Listing the license in the recipe is only to satisfy conda's handling of
the lic file and, bc these datasets have multiple licenses in multiple
files, that is just not practical.

This is an exception and we should be ok with the linter failure here.

**Comment by xylar** [comment]
Okay, I'll remove my last commit as soon as I get a chance.

File: recipes/proj-data/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+test:
+  commands:
+    - test -f $PREFIX/share/proj/ca_nrc_ntv1_can.tif
+    - echo -111 50 | cs2cs +proj=latlong +ellps=clrk66 +nadgrids=ca_nrc_ntv1_can.tif +to +proj=latlong +ellps=GRS80 +datum=NAD83
+    - echo -105 40 | cs2cs +init=epsg:4326 +to +init=epsg:2975
+
+about:
+  home: https://github.com/OSGeo/proj-data
+  license: Public domain | X/MIT | BSD-2/3/4-Clause | CC0 | CC-BY (v3.0 or later) | CC-BY-SA (v3.0 or later)
+  # Each grid has a *README.txt file with the license info, that is copied with
```
Comment (ocefpaf) [inline recipes/proj-data/meta.yaml]: Just leave this comment here. We don't need to list all the licenses below b/c that would be a huge mess.

**Comment by xylar** [comment]
Thanks @ocefpaf, I've removed the incorrect license.  @scopatz, will you be willing to accept this recipe as an exception to the requirement of having a single license per package, in light of @ocefpaf's explanation?

File: recipes/proj-data/meta.yaml (hunk 5 above)
Comment (scopatz) [reply recipes/proj-data/meta.yaml]: Sure, but if it is licensed under CC0, that is all we need to add.  We must have *some* license

Comment (ocefpaf) [reply recipes/proj-data/meta.yaml]: The licenses ARE in the package in the authors designed directories. We DO have all of them. This is does not copy the files over to the conda directory and that is a requirement that we created to cover the many cases where the file(s) did not even existed.

Comment (scopatz) [reply recipes/proj-data/meta.yaml]: `licesne_file` can be a list.  What is the problem with including at least one of them?

Comment (xylar) [reply recipes/proj-data/meta.yaml]: Each data set has its own license.  It is not that all these licenses apply to all the data.  If we picked an arbitrary data license for an arbitrary data set, this would give the false impression that this license applied to all the data sets.

Comment (ocefpaf) [reply recipes/proj-data/meta.yaml]: It will be misleading and wrong. Also, while the distribution of those data with multiple licenses is fine, listing a license or some licenses without mentioned which data they belong we will be infringing some of the license.

**Comment by xylar** [comment]
Thanks @ocefpaf!

**Comment by scopatz** [comment]
@ocefpaf It was wrong of you to merge this. I have a legitimate concern over the license file inclusion here.  Please revert

**Comment by ocefpaf** [comment]
> @ocefpaf It was wrong of you to merge this. I have a legitimate concern over the license file inclusion here. Please revert

I won't revert. All the required information is there. Your suggestion will cause license violation and/or non-compliance just to satisfy the linter in the recipe. That is a no-no IMO.

**Comment by xylar** [comment]
@ocefpaf and @scopatz, let's not escalate this unnecessarily.

@scopatz, assuming we were to either revert or address your concerns in the resulting feedstock, it remains unclear to me how we could provide a single license file that would be honest to the contents of the package.  The alternative seems absurd to me: each data set in its own package with a metapackage to contain them all (still needing a license, somehow!).  This package contains dozens of data sets if not hundreds, each with its own license.

**Comment by scopatz** [comment]
@xylar - I strongly feel that this recipe should be a meta-package and that there should be a separate output/package for each license. Not having a valid `license_file` entry breaks downstream APIs, and this package is likely to be removed from anything that does license checking

**Comment by scopatz** [comment]
@xylar - please fix this in the feedstock ASAP

**Comment by isuruf** [comment]
>  it remains unclear to me how we could provide a single license file that would be honest to the contents of the package.

It doesn't have to be. You can use a list of license files.

**Comment by xylar** [comment]
@isuruf, okay, I will try to do that.  Again, we're likely talking about dozens to hundreds of license files.  This is not a package I will ever use myself, I was just trying to do a service to the community here...  It's been a pretty frustrating experience.

**Comment by scopatz** [comment]
Yeah, I also suggested a list of license

**Comment by ocefpaf** [comment]
> It doesn't have to be. You can use a list of license files.

Listing cannot different between which data is under which license and that will cause issues @isuruf. This dataset is distributed like that in all the Linux world and brew. Add a license list just to comply with automated tools is counter-productive b/c there is nothing wrong with the package or the way it distributes its licenses.

TL;DR while creating multiple packages and then aggregating them in a meta-package may look like a way to solve this, remember, the meta-package must have a license, which one is it? How would that play out if the one chosen does is incompatible with another one and the users is checking on the meta-package?

The proj devs recommend the license to be in each dataset README exactly b/c of this.

This has became a tedious discussion between, let's give in to dumb automation vs doing the right thing.

**Comment by ocefpaf** [comment]
> Yeah, I also suggested a list of license

Again, that is wrong and will violate the licenses that does not play together.

**Comment by ocefpaf** [comment]
> @isuruf, okay, I will try to do that. Again, we're likely talking about dozens to hundreds of license files. This is not a package I will ever use myself, I was just trying to do a service to the community here... It's been a pretty frustrating experience.

If you do that you will have to remove some datasets. Debian did that a long time ago and reverted to this model exactly b/c of a similar discussion.

**Comment by isuruf** [comment]
> Debian did that a long time ago and reverted to this model exactly b/c of a similar discussion.

I think we are missing the context here. Do you have some links to this discussion? Is https://packages.debian.org/sid/proj-data the equivalent package in Debian?

**Comment by xylar** [comment]
Given the versioning, I doubt that package is quite the same thing but it might be a previous variant (perhaps what is in the [datumgrid-feedstock](https://github.com/conda-forge/datumgrid-feedstock)).  But the licensing would be quite similar.

**Comment by xylar** [comment]
Hmm, actually, given the size, I think it was a small subset of the data, perhaps small enough not to have the same licensing issues.

**Comment by ocefpaf** [comment]
> Hmm, actually, given the size, I think it was a small subset of the data, perhaps small enough not to have the same licensing issues.

Yep. It has been a while that I gave up on Debian but if they went down the road to list licenses they have to remove the ones that do not comply with each other and/or break up into small packages. That is against the recommendation upstream b/c it adds an unnecessary hurdle to install all the datasets.

**Comment by scopatz** [comment]
Meta-package licenses only cover the code in the recipe itself (usally BSD-3-Clause here).  No package on conda-forge has a license that covers it dependencies.

**Comment by isuruf** [comment]
> Yep. It has been a while that I gave up on Debian but if they went down the road to list licenses they have to remove the ones that do not comply with each other and/or break up into small packages

I have no idea what this about, but if they do not comply with each other, how can we distribute them together?

**Comment by hobu** [comment]
https://github.com/OSGeo/PROJ-data/#about-the-data-package

We are not letting grids into the PROJ CDN that do not have an OSI-approved license. Happy to defer to @conda-forge licensing protocol on how to break apart the package, but the PROJ-datumgrid package has the same issues but worse – some of those grid files had ambiguous terms, no license, etc.

**Comment by ocefpaf** [comment]
> I have no idea what this about, but if they do not comply with each other, how can we distribute them together?

Yep. Unless they have an explicit clause against that we can distribute them together, use them together, put them in the same environment together. What we cannot do is say in a conda-recipe metadata that a certain list of licenses covers all the datasets.

**Comment by hobu** [comment]
A full listing is available at https://github.com/OSGeo/PROJ-data/blob/master/copyright_and_licenses.csv

**Comment by isuruf** [comment]
I'm sorry, but I don't understand. Can you explain more the following comment?

> What we cannot do is say in a conda-recipe metadata that a certain list of licenses covers all the datasets.

If each dataset has a license, how is it not possible to say that the the files in the package are covered by one or more of the following list and we give a list?

**Comment by ocefpaf** [comment]
> A full listing is available at [OSGeo/PROJ-data:copyright_and_licenses.csv@`master`](https://github.com/OSGeo/PROJ-data/blob/master/copyright_and_licenses.csv) [](/OSGeo/PROJ-data/blob/HEAD@{2020-03-09T19:45:02Z}/copyright_and_licenses.csv)


@xylar while that is not a license file that has a 1-1 dataset-license correspondence. Adding that will help users find the license. (But anyone expecting a license file text will be disappointed.)

**Comment by scopatz** [comment]
Right, so this should be broken up into separate outputs.  Right now none of these have the correct license

**Comment by ocefpaf** [comment]
> Right, so this should be broken up into separate outputs. Right now none of these have the correct license

They all have the right license. The artificial conda metadata field used in simplistic automated systems that cannot find them. That is VERY different.

**Comment by scopatz** [comment]
@ocefpaf - I don't understand why this package gets to be special with respect to the curated quality standards that we have had in effect for years.

**Comment by snowman2** [comment]
Is this package worth all of this trouble with `projsync`? It has the `--all` option: https://proj.org/apps/projsync.html

**Comment by ocefpaf** [comment]
> @ocefpaf - I don't understand why this package gets to be special with respect to the curated quality standards that we have had in effect for years.

You mean from the many years where we never had the license_file field? Or the many package that infringe that by listing a single license when it should've been multiple?

By not listing license files we are simply telling the user: go look into the package. B/c they are all there. Again, you are arguing for conda-metadata just for the sake of automated tools.

**Comment by xylar** [comment]
```
Is this package worth all of this trouble with projsync? It has the --all option:
```
@snowman2, there are a number of machines that let anaconda.org through the firewall but not the server used by `projsync`, so for those cases it could be.

**Comment by xylar** [comment]
It took awhile for me to download the tarball again, but now that I have, I see that there are far fewer license files than I originally thought:
```
at_bev_README.txt            es_cat_icgc_README.txt
au_ga_README.txt             eur_nkg_README.txt
au_icsm_README.txt           fr_ign_README.txt
be_ign_README.txt            is_lmi_README.txt
ca_nrc_README.txt            nl_nsgi_README.txt
ca_que_mern_README.txt       nz_linz_README.txt
ch_swisstopo_README.txt      pt_dgt_README.txt
de_adv_README.txt            se_lantmateriet_README.txt
de_geosn_README.txt          uk_os_README.txt
de_lgl_bw_README.txt         us_nga_README.txt
de_lgvl_saarland_README.txt  us_noaa_README.txt
dk_sdfe_README.txt
```

I realize that @ocefpaf has strong objections but I personally don't see a reason these couldn't be listed in `license_file`.  Since each has a prefix that corresponds to the grids it applies to, it should be clear to everyone which license corresponds with which sets of grids, and I think someone would have to be being purposefully disingenuous to say they thought all licenses applied to all data sets.

**Comment by ocefpaf** [comment]
> I realize that @ocefpaf has strong objections

I'm not a lawyer and I'm echoing past decision to avoid mixing licenses that don't play with each other in the same listing. The CSV file has a 1-1 correspondence that is better than the license name there, I'd use that to satisfy automated tools. If you ensure that those license can be listed together without any ambiguity that they refer to its own dataset, then go for it.

PS: last time I had this discussion it was in the proj.4 era. A lot has changed since then and @snowman2 and @hobu are better equipped to say something about the licenses.

**Comment by xylar** [comment]
The CSV also works for me.  When I pulled it up, I didn't see the third column so I was concerned the info wasn't sufficient but now that I see the license column, it does look preferable to me.  I'll point to that as soon as the feedstock is created.

**Comment by scopatz** [comment]
> Again, you are arguing for conda-metadata just for the sake of automated tools.

Conda-forge would not exist without automated tools, so yes, I am arguing for automated tools

**Comment by ocefpaf** [comment]
> Conda-forge would not exist without automated tools, so yes, I am arguing for automated tools

As someone who is in conda-forge since day 0, wrote and re-wrote most of those tools. Writes grants to ensure its funding and participates in all the discussions about them... I know. I also know that blindly following them and filling bad metadata is worse. If there is any consequence about the tooling we should fix the tool.

Last but not least, this package is not alone in this exception. We have many like this one. That is the main reason why the blocked https://github.com/conda-forge/cfep/pull/23.

**Comment by isuruf** [comment]
Let's pause this discussion for now and try to come up with a fix in the feedstock.

**Comment by ocefpaf** [comment]
> Let's pause this discussion for now and try to come up with a fix in the feedstock.


Are you ok with the csv @isuruf?

**Comment by isuruf** [comment]
> Are you ok with the csv @isuruf?

Yes
